### PR TITLE
Refbox MQTT MPS communication

### PR DESCRIPTION
**Pull request by Karatekuh**
This pull request adds the possibility to the refbox to communicate with the MPS via MQTT.

### CODE REVIEW

File: src/libs/mps_comm/Makefile [hunk 1]
```diff
@@ -36,6 +38,13 @@ else
   WARN_TARGETS += warning_cpp17
 endif
 
+HAVE_MQTT = 1
```
Comment (TarikViehmann) [inline src/libs/mps_comm/Makefile:41]: Mmh the packages dont have pkg-config files, so a proper check is not possible.
We plan on migrating the refbox to cmake soon, but until then we should proably at least to a small sanity check whether the mqtt libs exist. something like this:

HAVE_MQTT = $(shell if [ -f /usr/include/mqtt/async_client.h ]; then echo 1; else echo 0; fi)

File: src/libs/mps_comm/Makefile [hunk 2]
```diff
@@ -20,13 +20,15 @@ include $(BUILDSYSDIR)/protobuf.mk
 include $(BUILDSYSDIR)/boost.mk
 include $(BASEDIR)/src/libs/mps_comm/freeopcua.mk
 
-LIBS_libmps_comm = stdc++ m llsfrbcore llsfrbconfig pthread
+LIBS_libmps_comm = stdc++ m llsfrbcore llsfrbconfig pthread paho-mqtt3cs paho-mqtt3c paho-mqtt3as paho-mqtt3a paho-mqttpp3
```
Comment (TarikViehmann) [inline src/libs/mps_comm/Makefile:23]: this should only be required when mqtt is built.

Suggested change here: c6f085ef986c958e95fd8e827abcc66dd9a706b6

File: src/libs/mps_comm/mqtt/base_station.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+namespace mps_comm {
+
+class MqttBaseStation : public virtual MqttMachine, public virtual BaseStation
+{
+public:
+	MqttBaseStation(const std::string &name,
+	                const std::string &ip,
+	                unsigned short     port,
+	                const std::string &log_path = "",
+	                ConnectionMode     mode     = PLC);
```
Comment (TarikViehmann) [inline src/libs/mps_comm/mqtt/base_station.h:40]: degfault mode for all mqtt machines should be MQTT not PLC

Suggested changes here: d3fbacb0dfbf3c5b7b66865753c310031df77272

File: src/libs/mps_comm/mqtt/mqtt_action_listener.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	void                            on_failure(const mqtt::token &tok) override;
+	void                            on_success(const mqtt::token &tok) override;
+
+public:
+	mqtt_action_listener(const std::string &name, std::shared_ptr<spdlog::logger> logger);
+	~mqtt_action_listener();
+};
+
+} // namespace mps_comm
+} // namespace llsfrb
```
Comment (TarikViehmann) [inline src/libs/mps_comm/mqtt/mqtt_action_listener.h:54]: bunch of missing newlines.

Suggested changes here: 19848670fd8cb49a4bb9e06fd629ea805f07f479

**Review by TarikViehmann — CHANGES_REQUESTED** [review]
Very nice work, I have a few minor suggestions, which I also have commits for that you can just cherry-pick if you like. The commits are in this branch: https://github.com/robocup-logistics/rcll-refbox/tree/common/mqtt-mps-communication

File: Dockerfile [hunk 5]
```diff
@@ -55,14 +61,27 @@ RUN shopt -s globstar; \
...
 RUN echo /usr/local/lib64 > /etc/ld.so.conf.d/local.conf && /sbin/ldconfig
-RUN dnf install -y --nodocs $(cat /requires.txt) && dnf clean all && rm /requires.txt
+RUN dnf install -y --nodocs 'dnf-command(copr)' && dnf -y copr enable tavie/paho-mqtt-cpp && dnf -y copr enable tavie/clips_protobuf &&\
+    dnf install -y --nodocs $(cat /requires.txt) && \ 
+    dnf clean all && \
+    cat /requires.txt && \
+    rm /requires.txt
 CMD ["llsf-refbox"]
+
+FROM builder as devcontainer
```
[the remark this answers is not in the document]
Comment (Karatekuh) [reply Dockerfile:83]: I had some issues with the devcontainer.Dockerfile, as the builder stage was not found and being unable to start the devcontainer correctly. I will have a look at it once more to make sure that this works.

File: src/libs/mps_comm/mqtt/mqtt_action_listener.h (hunk 4 above)
Comment (Karatekuh) [reply src/libs/mps_comm/mqtt/mqtt_action_listener.h]: cherry-picked the suggested changes

File: src/libs/mps_comm/mqtt/base_station.h (hunk 3 above)
Comment (Karatekuh) [reply src/libs/mps_comm/mqtt/base_station.h]: cherry-picked the suggested changes

File: src/libs/mps_comm/Makefile (hunk 2 above)
Comment (Karatekuh) [reply src/libs/mps_comm/Makefile]: cherry-picked the suggested changes

File: src/libs/mps_comm/Makefile (hunk 1 above)
Comment (Karatekuh) [reply src/libs/mps_comm/Makefile]: This change is also included in a cherry-pick so this thread is resolved.

File: Dockerfile [hunk 6]
```diff
@@ -55,14 +61,27 @@ RUN shopt -s globstar; \
...
+      dnf -y copr enable tavie/paho-mqtt-cpp && \
+      dnf install -y --nodocs paho-c paho-c-devel paho-mqtt-cpp paho-mqtt-cpp-devel 
 COPY --from=buildenv /buildenv/bin/* /usr/local/bin/
 COPY --from=buildenv /buildenv/lib/* /usr/local/lib64/
 COPY --from=buildenv /buildenv/src/games /usr/local/share/rcll-refbox/games
 COPY --from=buildenv /buildenv/src/msgs/*.proto /usr/local/share/rcll-refbox/msgs/
 COPY --from=buildenv /buildenv/src/libs/websocket/message_schemas/*.json /usr/local/share/rcll-refbox/libs/websocket/message_schemas/
 COPY --from=buildenv /buildenv/cfg /etc/rcll-refbox/
 COPY --from=buildenv /buildenv/requires.txt /
+RUN cat /requires.txt
```
[the remark this answers is not in the document]
Comment (Karatekuh) [reply Dockerfile]: Has been removed via the cherry-pick

File: Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (Karatekuh) [reply Dockerfile]: retried it, you have to build the builder image atleast once specifically with tagging then it should work going forward.

File: Dockerfile [hunk 7]
```diff
@@ -55,14 +61,27 @@ RUN shopt -s globstar; \
...
 RUN echo /usr/local/lib64 > /etc/ld.so.conf.d/local.conf && /sbin/ldconfig
-RUN dnf install -y --nodocs $(cat /requires.txt) && dnf clean all && rm /requires.txt
+RUN dnf install -y --nodocs 'dnf-command(copr)' && dnf -y copr enable tavie/paho-mqtt-cpp && dnf -y copr enable tavie/clips_protobuf &&\
+    dnf install -y --nodocs $(cat /requires.txt) && \ 
+    dnf clean all && \
```
[the remark this answers is not in the document]
Comment (Karatekuh) [reply Dockerfile]: Has been removed via the cherry-pick

**Review by TarikViehmann — APPROVED** [review]
Thank you very much!

**Pull request merged**